qiskit/algorithms/fidelities/base_fidelity.py

```diff
@@ -0,0 +1,253 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Base fidelity interface
+"""
+
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from typing import Sequence, Mapping
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import ParameterVector
+
+
+class BaseFidelity(ABC):
+    """
+    An interface to calculate fidelities (state overlaps) for pairs of
+    (parametrized) quantum circuits.
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        """Initializes the class to evaluate fidelities."""
+
+        self._circuits: Sequence[QuantumCircuit] = []
+        self._parameter_values: Sequence[Sequence[float]] = []
+
+        # use cache for preventing unnecessary circuit compositions
+        self._circuit_cache: Mapping[(int, int), QuantumCircuit] = {}
+
+        self._left_parameters = []
+        self._right_parameters = []
+
+    def _check_values_shape(
+        self,
+        values: Sequence[Sequence[float]] | None,
+        circuits: QuantumCircuit,
+        label: str | None = " ",
+    ) -> list[list[float]] | None:
+        """
+        Checks whether the passed values match the shape of the parameters
+        of the corresponding circuits and formats values to 2D list.
+
+        Args:
+            values: parameter values corresponding to the circuits to be checked
+            circuits: list of circuits to be checked
+            label: optional label to allow for circuit identification in error message
+
+        Returns:
+            Returns a 2D list if values match, `None` if no parameters are passed
+
+        Raises:
+            ValueError: if the number of parameter values doesn't match the number of
+                        circuit parameters
+        """
+
+        # Support ndarray
+        if isinstance(values, np.ndarray):
+            values = values.tolist()
+
+        if values is None:
+            for circuit in circuits:
+                if circuit.num_parameters != 0:
+                    raise ValueError(
+                        f"`values_{label}` cannot be `None` because circuit_{label} has "
+                        f"{circuit.num_parameters} free parameters."
+                    )
+            return None
+        else:
+            # ensure 2d list
+            if not isinstance(values, list):
+                values = [values]
+            if len(values) > 0 and not isinstance(values[0], list):
+                values = [values]
+            return values
+
+    def _check_qubits_mismatch(self, circuit_1: QuantumCircuit, circuit_2: QuantumCircuit) -> None:
```

Comment:
This is super minor, but I think `_check_qubits_match` (or even `_check_num_qubits_match`) would make more sense, as `_check_qubits_mismatch` leads me to assume that mismatching is the desired outcome. This might just be me though.

Reply:
haha it's true, thanks for the observation!